Relax requirements on key and value types

jrhemstad opened the pull request:
We previously required the key/value types to be arithmetic types. This is more strict than necessary.

This PR relaxes the requirements to key and value types that have unique object representations and are trivially copyable. This change required moving to C++17 to access the `has_unique_object_representations` trait. Added tests for user defined key and value types. 

Likewise, I had to add a new `bitwise_compare` function to use whenever comparing against a sentinel value (previously we just use `operator==` as it was guaranteed for arithmetic types. It is no longer guaranteed with these relaxations).

I also improved `cuco::pair` to allow for more generic construction from "pair like" things. This includes a `thrust::zip_iterator`.

jrhemstad committed:
Update cmake to require c++17.

jrhemstad committed:
Relax restrictions on key/value type to trivially copyable and unique object rep.

jrhemstad committed:
Add and use bitwise comparison for comparing against sentinel.

jrhemstad committed:
Update doc.

jrhemstad committed:
Remove requirement for arithmetic key types.

jrhemstad committed:
Allow construct cuco::pair from anything that is "pair like".

jrhemstad committed:
Add static_map tests using user defined, compound key/value types.

griwes commented on include/cuco/detail/bitwise_compare.cuh:
```diff
@@ -0,0 +1,77 @@
+struct bitwise_compare_impl {
+  __host__ __device__ static constexpr bool compare(char const* lhs, char const* rhs)
+  {
+    return memcmp(lhs, rhs, TypeSize) == 0;
```
`cuda_memcmp`?

jrhemstad committed:
Correct is_CAS_safe.

griwes commented on include/cuco/detail/pair.cuh:
```diff
@@ -41,10 +46,31 @@ constexpr std::size_t next_pow2(std::size_t v) noexcept {
+struct is_pair_like_impl : std::false_type {
+};
+
+template <typename T>
+struct is_pair_like_impl<T,
+                         std::void_t<decltype(thrust::get<0>(std::declval<T>())),
+                                     decltype(thrust::get<1>(std::declval<T>()))>>
+  : std::true_type {
+};
```
Is this meant to accept `thrust::tuple<int, float, bool>`?

jrhemstad replied:
No, but every attempt I made to use `std::tuple_size` would break something in Thrust, and `thrust::tuple_size` would break something in `std::`. I suppose I could SFINAE on both of them.

griwes reviewed (commented):
The approach looks good to me, see comments below.

jrhemstad committed:
Remove float from value_pair.

has_unique_object_representations<float>
is false.

jrhemstad committed:
Use separate pair_like traits for std and thrust.

std and thrust tuple machinery are very different.
Therefore, had to specialize for each separately.

jrhemstad committed:
Use cuda_memcmp.
